src/task.c

```diff
@@ -133,23 +168,86 @@ task_register_hook(task_hook newhook)
    return STATUS_OK;
 }
 
-// Register the calling thread and allocate scratch space for it
+/*
+ * task_register_thread(): Register a new thread with the task system.
+ *
+ * Registration implies:
+ *  - Acquire a new thread ID (index) for this to-be-active thread
+ *  - Allocate scratch space for this thread, and track this space..
+ */
 void
-task_register_this_thread(task_system *ts, uint64 scratch_size)
+task_register_thread(task_system *ts,
+                     uint64       scratch_size,
+                     const char  *file,
+                     const int    lineno,
+                     const char  *func)
 {
+   threadid thread_tid = platform_get_tid();
+   /*
+    * For all newly created threads, the thread-index will be 0 before the hook
+    * function, run below, generates this thread's index. We do not want to
+    * register a thread twice; otherwise, it will simply burn up an index and
+    * possibly allocated scratch space.
+    * As threads generally will require scratch space to be allocated, an
+    * attempt to re-register a thread should find scratch space already
+    * allocated. Flag that.
+    */
+   platform_assert((thread_tid == 0)
```

Comment:
Added ... Useful assertion. Found and tripped on incorrect usages in test code that was resulting in nested register thread for the same task.